File: task3.py
from lib.quick_sort import quick_sort

def solution(L: int, N: int, points: list[int]) -> int:
    '''
    На прямой располагается 1 ≤ N ≤ 10000 точек с целочисленными координатами –10**9
    ≤ Vi ≤ 10**9. Каждой из точек разрешается сделать ровно одно движение (танцевальное па) в
    любом направлении на расстояние не больше 0 ≤ L ≤ 10**8 и остановиться на другой позиции.
    Какое минимальное количество точек может остаться на прямой после окончания танца
    (все точки после танца, оказывающиеся на одной позиции, сливаются в одну)?
    Args:
        L (int): max distance 0 <= L <= 10**8
        N (int): points count 1 <= M <= 10**5
        points (list[int]): list of points with positions -10**9 <= points[i] <= 10**9
    Returns:
        int: MinimalNumberOfPoints
    '''
    quick_sort(points)

    count = 1

    current_end = points[0] + L

    for i in range(1, N):
        if points[i] >= current_end:
            count += 1
        current_end = points[i] + L
    
    return count


def main():
    import sys
    [L, N] = list(map(int, sys.stdin.readline().split()))
    points = list(map(int, sys.stdin.readline().split()))
    
    result = solution(L, N, points)
    print(result)

if __name__ == "__main__":
    main()

File: task3.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class task3 {
    
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String[] data = reader.readLine().trim().split(" ");
        int L = Integer.parseInt(data[0]);
        int N = Integer.parseInt(data[1]);

        String[] pointsStr = reader.readLine().trim().split(" ");
        
        int[] points = new int[N];
        for (int i = 0; i < N; i++) {
            points[i] = Integer.parseInt(pointsStr[i]);
        }
        
        int result = solution(L, N, points);
        System.out.println(result);
    }
    
    public static int solution(int L, int N, int[] points) {
        QuickSort.quickSort(points);
        
        int count = 1;
        long currentEnd = (long) points[0] + L;
        
        for (int i = 1; i < N; i++) {
            if (points[i] >= currentEnd) {
                count++;
            }
            currentEnd = (long) points[i] + L;
        }
        
        return count;
    }
}
